lib/teleterm/services/connectmycomputer/connectmycomputer.go

```diff
@@ -0,0 +1,280 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connectmycomputer
+
+import (
+	"context"
+	"fmt"
+	"os/user"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/teleterm/clusters"
+)
+
+type RoleSetup struct {
+	cfg *RoleSetupConfig
+}
+
+func NewRoleSetup(cfg *RoleSetupConfig) (*RoleSetup, error) {
+	err := cfg.CheckAndSetDefaults()
+	if err != nil {
+		return nil, err
+	}
+
+	return &RoleSetup{cfg: cfg}, nil
+}
+
+type RoleSetupResult struct {
+	CertsReloaded bool
+}
+
+// Run ensures that at the end of its execution the user has their own individual Connect My
+// Computer role and that the role includes the current system username in allowed logins.
+//
+// If the role list of the user got updated, the return value has CertsReloaded set to true.
+func (s *RoleSetup) Run(ctx context.Context, accessAndIdentity AccessAndIdentity, certManager CertManager, cluster *clusters.Cluster) (RoleSetupResult, error) {
+	noCertsReloaded := RoleSetupResult{}
+	if !cluster.URI.IsRoot() {
+		return noCertsReloaded, trace.BadParameter("Connect My Computer works only with root clusters")
+	}
+
+	// Do not use GetCurrentUser – it returns the current view of the user given the certs, not merely
+	// the resource from the backend. This means that GetCurrentUser includes any roles granted
+	// through access requests. We don't want that since we're later going to update the user.
+	clusterUser, err := accessAndIdentity.GetUser(cluster.GetLoggedInUser().Name, false /* withSecrets */)
+	if err != nil {
+		return noCertsReloaded, trace.Wrap(err)
+	}
+
+	isLocalUser := clusterUser.GetCreatedBy().Connector == nil
+	if !isLocalUser {
+		return noCertsReloaded,
+			trace.BadParameter("Connect My Computer works only with local users, the user %v was created by %v connector",
+				clusterUser.GetName(), clusterUser.GetCreatedBy().Connector.Type)
+	}
+
+	roleName := fmt.Sprintf("%v%v", teleport.ConnectMyComputerRoleNamePrefix, clusterUser.GetName())
+
+	doesRoleExist := true
+	existingRole, err := accessAndIdentity.GetRole(ctx, roleName)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			doesRoleExist = false
+		} else {
+			return noCertsReloaded, trace.Wrap(err)
+		}
+	}
+
+	systemUser, err := user.Current()
+	if err != nil {
+		return noCertsReloaded, trace.Wrap(err)
+	}
+
+	if !doesRoleExist {
+		s.cfg.Log.Infof("Creating the role %v.", roleName)
+
+		role, err := types.NewRole(roleName, types.RoleSpecV6{
+			Allow: types.RoleConditions{
+				NodeLabels: types.Labels{
+					types.ConnectMyComputerNodeOwnerLabel: []string{clusterUser.GetName()},
+				},
+				Logins: []string{systemUser.Username},
+			},
+		})
```

Comment:
Maybe this is unlikely, but I'm imagining a team with hundreds of users who all want to "add my computer". This will result in hundreds of very similar roles, which is just unecessary, and will interact poorly when admins prefer to manage roles via terraform or the k8s operator.

Could we get away with a single templated role? We'd have to add a couple traits to the user in addition to the role. Something like this:

```go
		role, err := types.NewRole(roleName, types.RoleSpecV6{
			Allow: types.RoleConditions{
				NodeLabels: types.Labels{
					types.ConnectMyComputerNodeOwnerLabel: []string{`{{external["teleport.internal/connect-my-computer-username"]}}`},
				},
				Logins: []string{`{{external["teleport.internal/connect-my-computer-logins"]}}`},
			},
		})
```

Reply:
just wanted to float this idea to get your take on it by the way, if you want to just go ahead with what you have here and in the RFD I'm alright with that, I think my imagined scenario is probably unlikely

Reply:
This is an excellent idea that I wish we thought about earlier during the RFD stage.

I think at the moment I'd prefer to focus on other tasks to get this feature out the door. Connect My Computer is going to be initially limited to Team plan clusters only. So as you said, a scenario with tons of extra roles is unlikely.

If we ever plan on expanding this feature, we'll probably have to [integrate it with the backend more tightly](https://github.com/gravitational/teleport/blob/c519c51378f7d4d9a754dcc6edb5b7e8bf56930b/rfd/0133-connect-my-computer.md#access-to-the-node), possibly removing the need for a role like this altogether:

> Accordingly, we assume that the MVP of Connect My Computer is going to be used in highly trusted
environments. Most further developments of this feature will require reinforcing the security of it.
[Mike Jensen composed a list of potential ideas](https://github.com/gravitational/teleport/pull/27815#discussion_r1238756710) as to how this could be achieved.

But I'm going to capture this in our tracking issue. If we have enough time before the release, I'll try to address this.